### PR TITLE
Extract korlibs-datastructure

### PR DESCRIPTION
**Pull request by holloszaboakos**
Partial work on https://github.com/korlibs/korge/issues/2040

### CODE REVIEW

**Review by soywiz — CHANGES_REQUESTED** [review]
Hi @holloszaboakos thanks for the PR, let me comment a few adjustments that should be performed:

## `korlibs.platform` package

I have checked where it is used. It seems that it is used only in a class in the JS target, and in tests:

Screenshot 2024-01-03 at 11 23 18

This is used mainly to detect runtimes etc.

Let's create a `korlibs-platform` module:

It can be included in tests like this:

```
commonTestApi(project(":korlibs-platform"))
```

Let's keep the `korlibs.datastructure.event` in the `korge-foundation` for now. That way the main source set can be kept without depending on `korlibs-platform`. As it is not required.

## Buffer, UShortArrayInt, UByteArrayInt

They are not used, and since it is internal the arraycopy variants using them should be removed. You can search for references and delete the unused ones.

## Only `korlibs.platform`

Please, move the remaining stuff from other packages: InternalMath and InternalMemory (arraycopy etc.) to the `korlibs.platform.internal` package to clearly state this is internal utilities to that package and module.

**Comment by holloszaboakos** [comment]
@soywiz Hi! I am not 100% sure what are you asking. We are going to split the foundation further. Therefor this issue will be completly resolved when platfor fets its own module. Are you sure we should leave classes temporally in foundation? I would trather do the whole foundation split in one PR instead.

**Comment by soywiz** [comment]
> @soywiz Hi! I am not 100% sure what are you asking. We are going to split the foundation further. Therefor this issue will be completly resolved when platform gets its own module. Are you sure we should leave classes temporally in foundation? I would rather do the whole foundation split in one PR instead.

The EventLoop stuff might be moved later to korge itself where it is used keeping its package. Not directly a datastructure. So yeah, let's keep it in the `korge-foundation` for now. I don't think it makes sense to include it in `korlibs-datastructure`.

Reviewing the whole foundation split would be a lot of work and riskier if some changes happen in the meantime. I would prefer doing one module at a time.

Let's focus on one thing at a time, small steps:

* The `korlibs-platform` doesn't have dependencies to anything else. That could be the best first split. datastructures only use it in tests.
* Place only the functions that are used in the `korlibs.datastructure.internal` package instead of `korlibs.mathinternal` etc. it should be a few functions like ilog2, some arraycopy signatures and not too much else.

**Comment by soywiz** [comment]
Please, to avoid having duplicate code that is unused, remove Buffer, Buffer-related arraycopy variants and UByte/UShortArrayInt:

Screenshot 2024-01-03 at 14 30 21

**Comment by soywiz** [comment]
And all the functions that appear grey (unused):

Screenshot 2024-01-03 at 14 32 18

The rest that are still used, should be marked as `internal` to not be exposed.

**Comment by holloszaboakos** [comment]
@soywiz  I am going to make the deletion of not needed features as a last step after everithing build correctly and all tests succeed.

File: settings.gradle.kts [hunk 1]
```diff
@@ -41,3 +46,4 @@ if (System.getenv("DISABLE_SANDBOX") != "true") {
 if (!inCI || System.getenv("ENABLE_BENCHMARKS") == "true") {
     include(":korge-benchmarks")
 }
+include("korlibs-platform")
```
Comment (soywiz) [inline settings.gradle.kts:49]: This line is duplicated

File: korge-core/build.gradle.kts [hunk 2]
```diff
@@ -5,6 +5,8 @@ dependencies {
         //add("androidMainApi", "androidx.javascriptengine:javascriptengine:1.0.0-alpha05")
         //add("androidMainApi", "com.google.guava:guava:31.0.1-android")
     }
+    commonMainApi(project(":korlibs-datastructure"))
+    commonMainApi(project(":korlibs-platform"))
```
Comment (soywiz) [inline korge-core/build.gradle.kts:9]: This should be removed. I mean, the idea is that those artifacts are transitively resolved via `korge-foundation` that is already included.

File: korlibs-datastructure/src/korlibs/datastructure/_Datastructure_event.kt [hunk 3]
```diff
@@ -88,7 +87,7 @@ open class SyncEventLoop(
             if (running) {
                 timedTasks.add(task)
             } else {
-                Console.warn("WARNING: QUEUED TASK time=$time interval=$interval without running")
+                //TODO: add back logger Console.warn("WARNING: QUEUED TASK time=$time interval=$interval without running")
```
Comment (soywiz) [inline korlibs-datastructure/src/korlibs/datastructure/_Datastructure_event.kt:90]: This class/package should be kept in the korge-foundation for now, and there the logger package should be available.

File: settings.gradle.kts [hunk 4]
```diff
@@ -13,6 +13,9 @@ pluginManagement {
         maven { url = uri("https://oss.sonatype.org/content/repositories/snapshots/") }
     }
 }
+plugins {
+    id("org.gradle.toolchains.foojay-resolver-convention") version "0.5.0"
+}
```
[the remark this answers is not in the document]
Comment (holloszaboakos) [reply settings.gradle.kts]: I did not add it manually. It must have been intelij. Resolved.

File: korlibs-platform/src@android/korlibs/platform/Current.android.kt [hunk 5]
```diff
@@ -1,8 +1,8 @@
 package korlibs.platform
 
 import android.os.*
-import korlibs.korge_foundation.*
-import korlibs.memory.*
+//import korlibs.korge_foundation.*
```
[the remark this answers is not in the document]
Comment (holloszaboakos) [reply korlibs-platform/src@android/korlibs/platform/Current.android.kt:4]: Unresolved reference: korge_foundation
Unresolved reference: memory

[the remark this answers is not in the document]
Comment (holloszaboakos) [reply korlibs-platform/src@android/korlibs/platform/Current.android.kt:4]: something is not generating properly

**Comment by holloszaboakos** [comment]
The last thing to do is moving the event stuff back to foundation

**Comment by holloszaboakos** [comment]
@soywiz It is complete

File: buildSrc/src/main/kotlin/korlibs/korge/gradle/targets/android/AndroidDirect.kt [hunk 6]
```diff
@@ -84,9 +82,7 @@ fun Project.configureAndroidDirect(projectType: ProjectType, isKorge: Boolean) {
         }
 
         buildFeatures.apply {
-            if (project.name == "korge-foundation") {
-                buildConfig = true
-            }
+            buildConfig = true
```
Comment (soywiz) [inline buildSrc/src/main/kotlin/korlibs/korge/gradle/targets/android/AndroidDirect.kt:85]: Now that we know how it works, can you restore this if with `korlibs-platform`?

File: korge-foundation/src@jvmAndroid/korlibs/event/_Datastructure_event.jvmAndroid.kt [hunk 7]
```diff
@@ -1,6 +1,6 @@
 @file:Suppress("PackageDirectoryMismatch")
 
-package korlibs.datastructure.event
+package korlibs.event
```
Comment (soywiz) [inline korge-foundation/src@jvmAndroid/korlibs/event/_Datastructure_event.jvmAndroid.kt:3]: This is wrong. It should be `korlibs.datastructure.event` even if it is in the `korlibs-foundation` module. Otherwise we would be breaking compatibility. And we don't want that in this PR

**Comment by soywiz** [comment]
I have noticed that there are a few files that are deleted instead of moved:

image

Then re-added, any chance they could appear here as moved so it is easier to review?

File: korlibs-datastructure/test/korlibs/datastructure/ByteArrayDequeTest.kt [hunk 8]
```diff
@@ -26,7 +25,9 @@ class ByteArrayDequeTest {
         data.write(byteArrayOf(0x99.toByte(), 0x88.toByte(), 0x33, 0x44, 0x55, 0x66, 0xAA.toByte()))
         assertEquals(8, data.availableRead)
         assertEquals(0, data.availableWriteWithoutAllocating)
-        assertEquals(listOf(0x7F, 0x99, 0x88, 0x33, 0x44, 0x55, 0x66, 0xAA), data.readBytesUpTo(8).map { it.toInt() and 0xFF })
+        assertEquals(
+            listOf(0x7F, 0x99, 0x88, 0x33, 0x44, 0x55, 0x66, 0xAA),
+            data.readBytesUpTo(8).map { it.toInt() and 0xFF })
```
Comment (soywiz) [inline korlibs-datastructure/test/korlibs/datastructure/ByteArrayDequeTest.kt:30]: Why this change? Can we have moved files without changing imports or reformatting? We can do reformatting / optimizing imports in a separate PR if required.

File: korlibs-platform/src/korlibs/platform/Current.kt [hunk 9]
```diff
@@ -4,7 +4,7 @@ internal expect val currentOs: Os
 internal expect val currentRuntime: Runtime
 internal expect val currentArch: Arch
 internal expect val currentIsDebug: Boolean
-internal expect val currentIsLittleEndian: Boolean
+expect val currentIsLittleEndian: Boolean
```
Comment (soywiz) [inline korlibs-platform/src/korlibs/platform/Current.kt:7]: This shouldn't be used.

File: korlibs-datastructure/test/korlibs/datastructure/ByteArrayDequeTest.kt (hunk 8 above)
Comment (holloszaboakos) [reply korlibs-datastructure/test/korlibs/datastructure/ByteArrayDequeTest.kt:30]: I always end the PR by formatting. Turning back now is too much effort. If you prefer other formatting than I suggest edding automated formatting so it happens when someone pushes the commit.

**Comment by holloszaboakos** [comment]
> I have noticed that there are a few files that are deleted instead of moved:
> 
> image
> Then re-added, any chance they could appear here as moved so it is easier to review?

These files were only moved, not changed. They should be fine. The only event test is moved back to foundation.

**Comment by soywiz** [comment]
Superseded by:
* https://github.com/korlibs/korge/pull/2122
* https://github.com/korlibs/korge/pull/2123
